Add routing

td-extreme opened the pull request.

td-extreme committed:
Implement a better router

Removed the old router. The new router has a function that takes in a
method path and a list of routes. It seraches the list of routes for one
that matches the method and path and then returns the handler for that
route.

td-extreme committed:
Changed handler and core to use the new router

td-extreme committed:
Moved routes to sperate ns

This is to seperate the creation of routes from main.

td-extreme committed:
Renamed test descriptions

mikeknep commented on src/httpserver/routing/router.clj:
```diff
@@ -0,0 +1,15 @@
+  )
+
+(defn add-routes [server list-of-routes main-handler]
+  (doseq [route list-of-routes]
```
I think you could use `run!` here ([documentation](https://clojuredocs.org/clojure.core/run!))

mikeknep commented on src/httpserver/routing/router.clj:
```diff
@@ -0,0 +1,15 @@
+
+(defn get-handler [request routes]
+  (:handler (find-route (:method request) (:path request) routes))
+  )
```
newline

mikeknep commented on src/httpserver/handlers/handler.clj:
```diff
@@ -1,24 +1,27 @@
 
 (defn build-request [httpRequest]
-  (hash-map :path (.path httpRequest) :method (.method httpRequest) :headers (.headers httpRequest) :body (.body httpRequest)))
+  {:path (.path httpRequest) :method (.method httpRequest) :headers (.headers httpRequest) :body (.body httpRequest)})
```
Since commas are optional in Clojure, you'll often see a long line like this broken into newlines for key/value pairs, i.e.

``` clojure
{:path (.path httpRequest)
 :method (.method httpRequest)
 :headers (.headers httpRequest)
 :body (.body httpRequest)})
```

Just helps for readability

mikeknep commented:
I'm on the fence about `routes/` and `routing/` being separate directories. @dadarek any opinion?

mikeknep commented on spec/httpserver/handlers/handler_spec.clj:
```diff
@@ -1,13 +1,27 @@
-    (let [httpRequest (HttpRequest. "GET /coffee HTTP/1.1" (java.util.HashMap. {"this" "that"}))
-          main-handler (handler/new-handler)
-          httpResponse (.generateResponse main-handler httpRequest)]
+      (let [test-list (list {:method "GET" :path "/coffee" :handler teapot/call}
```
Rather than `(list {} {} {})`, you could use `'({} {} {})` (a true list) or `[{} {} {}]` (a vector). You'll probably see the vector more often—lists are most often employed when working with macros. See [this](http://stackoverflow.com/questions/11504236/as-a-data-container-what-are-the-main-differences-between-vector-and-list) and [this](http://stackoverflow.com/questions/1147975/in-clojure-when-should-i-use-a-vector-over-a-list-and-the-other-way-around) for descriptions of the differences between lists and vectors.

Also, just FYI, the quote in the example above is literally called "quote", and is not a typo. As Colin describes in his Clojure Macros book, you can think of it as meaning "literally". So, `(a b c)` = "call a with b and c as args", while `'(a b c)` = "literally, a b c (in a list)"

This is in a few places throughout the code. I'd recommend converting all `(list a b c)` 's to `[a b c]`

mikeknep commented on spec/httpserver/handlers/handler_spec.clj:
```diff
@@ -1,13 +1,27 @@
-          httpResponse (.generateResponse main-handler httpRequest)]
+      (let [test-list (list {:method "GET" :path "/coffee" :handler teapot/call}
+                            {:method "GET" :path "/test"}
+                            {:method "PUT" :path "/test"})
```
These two routes are unnecessary for the test. I think you can just remove them and just have `test-list` be a one-element vector

mikeknep commented on spec/httpserver/routing/router_spec.clj:
```diff
@@ -0,0 +1,30 @@
+  (it "returns false if neither the method nor path do not match the ones it should be looking for"
+    (should= false (router/check-route "PUT" "/" {:method "GET" :path "/path"})))
+  (it "returns true if the method and path match the ones it should be looking for"
+    (should= true (router/check-route "GET" "/path" {:method "GET" :path "/path"}))))
```
This test is good evidence of TDD-ing this feature, but I think `router/check-route` should ultimately be private. This is a situation where you would write out the specs as you test-drive the feature, then write tests for the "higher up" function (`router/find-route`), then go back and make this function private and delete these tests.

In fact, `router/find-route` might warrant being private too—it only gets called by `router/get-handler`, correct? So maybe you should only be testing this namespace via `get-handler`...

PS: to make a function private, define it with a hyphen: `(defn- this-is-private ...)`

td-extreme replied:
I had not even thought about public / private in clojure. But now that you brought it up, yes those should both be private. Since they are just helper functions anyway.

mikeknep commented on spec/httpserver/routing/router_spec.clj:
```diff
@@ -0,0 +1,30 @@
+    (let [test-list (list {:method "GET" :path "/" :handler "GET / hanlder"}
+                          {:method "GET" :path "/test" :handler "GET /test handler"})
+          request {:method "GET" :path "/test"}]
+      (should= "GET /test handler" (router/get-handler request test-list)))))
```
This confused me at first—"Why does `get-handler` return a string? But now I see you're mocking it out above. Solid approach, I would just recommend being clearer about the mock; maybe something like `{:method "GET" :path "/test" :handler :mock-test-handler}`

mikeknep commented on src/httpserver/routing/router.clj:
```diff
@@ -0,0 +1,15 @@
+  (some #(when (check-route method path %) %) routes))
+
+(defn get-handler [request routes]
+  (:handler (find-route (:method request) (:path request) routes))
```
This reads a little long for a one-liner. On the other hand, a `let` form seems a little heavy-handed...

Alright, it's time to learn about destructuring!

If you know the shape of an argument to your function, you can destructure it into its component parts. See if you can figure out what's going on in the example below—_don't implement it in your code until you understand it!_ There are lots of tutorials/blog posts/examples on the internet about destructuring, and I'm sure some apprentices would be willing to help explain it—it's a fun concept :)

``` clojure
(defn get-handler [{:keys [method path]} routes]
  (:handler (find-route method path routes)))
```

mikeknep commented on src/httpserver/handlers/handler.clj:
```diff
@@ -1,24 +1,27 @@
 
-(defn new-handler []
+(defn new-handler [list-of-routes]
+  (def routes list-of-routes)
```
If you want to use state, rather than re-`def`-ing, you should use an "atom". However, I don't think you need to do that here. I **think** you could do something like this:

``` clojure
(deftype main-handler [routes]
  com.td.HttpServer.IHandler
  (generateResponse [this httpRequest]
    ;; pretty sure you have access to routes here, since the handler was constructed with it
    ...))

(defn new-handler [list-of-routes]
  (main-handler. list-of-routes))
```

Now you're back to being stateless! Hooray!

mikeknep replied:
Followup on this: Geoff Shannon says that generally speaking you should use `defrecord` over `deftype`. I'm actually pretty rusty on the difference myself—Geoff says that `deftype` is for if you want to implement your own data structures.

Maybe worth researching and blogging?

Also, idiomatically the name should be Java style since they're actually creating Java classes/objects, i.e. `(defrecord MainHandler)` and `(MainHandler. list-of-routes)`

td-extreme committed:
Made helper functions private

Removed tests for functions that are now private.

td-extreme committed:
Removed seting of state of list-of-routes

list-of-routes is now passed into Handler during inti

Also changed the name of handler to Java style naming as it is
technically a java class.

td-extreme committed:
Fixed formating

td-extreme committed:
Moved routes into routing folder

mikeknep commented on src/httpserver/routing/router.clj:
```diff
@@ -0,0 +1,13 @@
+  (:handler (find-route method path routes)))
+
+(defn add-routes [server list-of-routes main-handler]
+  (run! #(.addRoute server (:method %) (:path %) main-handler) list-of-routes))
```
This is right on the fence for where I'd make this a private function, `add-route`, in terms of complexity.

td-extreme committed:
Chagned core to use routes in routing dir

td-extreme committed:
Made helper function add-route
